sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,539 @@
+# Dynamic linking of device code
+
+This document describes purpose and design of dynamic linking of device code
+feature.
+
+## Background
+Sometimes users want to link device code dynamically at run time. One possible
+use case for such linkage - providing device functions via shared libraries.
+The example below shows how device function `LibDeviceFunc` can be dynamically
+linked to a SYCL app:
+```
+// app.cpp
+SYCL_EXTERNAL int LibDeviceFunc(int i);
+class KernelName;
+/* ... */
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for<KernelName>(/* ... */ [=](sycl::item i) {
+  out[i] = LibDeviceFunc(i);
+}); /* ... */
+std::cout << out[i] << “ “;
+
+// lib.cpp
+int SYCL_EXTERNAL LibDeviceFunc(int i) {
+  return i * 2;
+}
+
+// Commands
+clang++ -fsycl lib.cpp -shared -o helpers.so
+clang++ -fsycl app.cpp -lhelpers -o a.out
```

Comment:
> Isn't this going to create host library, while you are trying to illustrate linking of "device code"?

This is going to create a "fat" library, which contains both host and device code. Host part will be linked automatically by regular C++ toolchain and system linker, while for device part we have to introduce new functionality, which is described in this design document.

> Or, maybe SYCL_EXTERNAL cases build for device?

Yes, `SYCL_EXTERNAL` means that this function will be available in device code even though it wasn't called from any SYCL kernel function.

> Which device then?

As usual with non-AOT workflow we do not specify the target device, because we don't know it in advance. Our compiler just allows to enable offloading in general by using SPIR-V intermediate representation

Reply:
OK. Can this be clarified in the writing please?

Reply:
Sure. Please check the paragraph that comes just after commands and let me know if it is enough to clarify.